### PR TITLE
#387 #399 #401 Sheet Model Improvements

### PR DESCRIPTION
**Pull request by Erik9519**
#### What have I done
Support for custom sheet color and support for choosing whether the sheet should be positioned at the center, on top or underneath the given route curve.
Fixed sheet model stretching.
Support for flipping sheet model. (Bottom side becoming top side and vice-versa)

#### Testing
Added Unit testing for Sheet Model moveTo

#### Addresses SRD section(s)
Sheet Model

### CODE REVIEW

File: src/app/core/models/sheet-model.ts [hunk 1]
```diff
@@ -1,16 +1,20 @@
 import * as THREE from 'three';
-import { SheetTextureSide } from './sheet-texture-side.enum';
-import { CurvePoint } from './curve-point';
-import { RouteCurve } from './route-curve';
+import {SheetTextureSide} from './sheet-texture-side.enum';
```
[the remark this answers is not in the document]
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Fixed

File: src/app/core/models/sheet-model.ts [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
+import {SheetCurveAlignment} from './sheet-curve-allignment.enum';
 
 export class SheetModel {
 
     /**
-     * Color of the sheet
+     * Color of the sheet texture
      */
-    private readonly COLOR = 0xffffff;
     private readonly TEXTURE_TEXT_COLOR = 'black';
```
[the remark this answers is not in the document]
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Addressed

File: src/app/core/models/sheet-model.ts [hunk 3]
```diff
@@ -60,39 +79,53 @@ export class SheetModel {
...
         // Mesh textures
-        const topTexture: THREE.Texture = this.generateTexture('x');
-        const bottomTexture: THREE.Texture = this.generateTexture('o');
-        const materials: THREE.Material[] = new Array(Object.keys(SheetTextureSide).length);
-        materials[SheetTextureSide.TOP] = new THREE.MeshPhongMaterial({
-            color: this.COLOR,              // Color of the Sheet
+        const topTexture: THREE.Texture = this.generateTexture('x', this.TEXTURE_TEXT_COLOR);
+        const bottomTexture: THREE.Texture = this.generateTexture('o', this.TEXTURE_TEXT_COLOR);
+        const material: THREE.Material[] = new Array(Object.keys(SheetTextureSide).length);
+        material[SheetTextureSide.TOP] = new THREE.MeshPhongMaterial({
```
[the remark this answers is not in the document]
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Created function

File: src/app/core/models/sheet-model.ts [hunk 4]
```diff
@@ -39,18 +43,33 @@ export class SheetModel {
...
+        height: number,
+        thickness: number,
+        segments: number,
+        color: number,
+        leadingEdgeColor: number
+    ) {
         this.width = width;
         this.height = height;
         this.thickness = thickness;
         this.segments = segments;
```
[the remark this answers is not in the document]
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Renamed numberOfParts

File: src/app/core/models/sheet-model.ts [hunk 5]
```diff
@@ -113,46 +156,82 @@ export class SheetModel {
...
+                -curvePoints[i].getTangent().x
             ).normalize();
-            const dirThickness: THREE.Vector3 = normal.multiplyScalar(this.thickness / 2);
             // Pre-Compute X and Y
-            const position: THREE.Vector3 = new THREE.Vector3(
+            const position: THREE.Vector2 = new THREE.Vector2(
                 curvePoints[i].getPoint().x - curvePoints[0].getPoint().x,
-                curvePoints[i].getPoint().y - curvePoints[0].getPoint().y,
-                0);
+                curvePoints[i].getPoint().y - curvePoints[0].getPoint().y);
```
[the remark this answers is not in the document]
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Fixed

File: src/app/workspace/viewport/viewport.component.ts [hunk 6]
```diff
@@ -233,13 +243,13 @@ export class ViewportComponent implements AfterViewInit {
         this.xGrid = new THREE.GridHelper(
             this.groundSize,
             this.groundSize,
-            0x000ff,
+            0x333333,
```
[the remark this answers is not in the document]
Comment (Erik9519) [reply src/app/workspace/viewport/viewport.component.ts]: Done

File: src/app/core/models/sheet-model.spec.ts [hunk 7]
```diff
@@ -1,10 +1,19 @@
-import { SheetModel } from './sheet-model';
+import {SheetModel} from './sheet-model';
 import * as THREE from 'three';
-import { SheetTextureSide } from './sheet-texture-side.enum';
+import {SheetTextureSide} from './sheet-texture-side.enum';
```
Comment (Dacaspex) [inline src/app/core/models/sheet-model.spec.ts]: Code style : spaces between brackets

File: src/app/core/models/sheet-model.ts [hunk 8]
```diff
@@ -2,15 +2,19 @@ import * as THREE from 'three';
 import { SheetTextureSide } from './sheet-texture-side.enum';
 import { CurvePoint } from './curve-point';
 import { RouteCurve } from './route-curve';
+import { SheetCurveAlignment } from './sheet-curve-allignment.enum';
 
 export class SheetModel {
 
     /**
-     * Color of the sheet
+     * Color of the sheet texture
```
Comment (Dacaspex) [inline src/app/core/models/sheet-model.ts]: Color of the sheet **text** texture, right?

File: src/app/core/models/sheet-model.ts [hunk 9]
```diff
@@ -347,7 +426,8 @@ export class SheetModel {
         ctx.fillStyle = 'white';
         ctx.fill();
         // Draw the txt
-        ctx.fillStyle = this.TEXTURE_TEXT_COLOR;
+        // tslint:disable-next-line:no-bitwise
+        ctx.fillStyle = '#' + ('00000' + (txtColor | 0).toString(16)).substr(-6);
```
Comment (Dacaspex) [inline src/app/core/models/sheet-model.ts]: Superlative of academic hack

File: src/app/core/models/sheet-model.spec.ts (hunk 7 above)
Comment (Erik9519) [reply src/app/core/models/sheet-model.spec.ts]: Addressed

File: src/app/core/models/sheet-model.ts (hunk 8 above)
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Yes

File: src/app/core/models/sheet-model.ts (hunk 9 above)
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Hurrah for hacks. it's gone

**Pull request merged**